pallets/corporate-actions/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+    fn set_did_withholding_tax() -> Weight;
+}
+
+/// The module's configuration trait.
+pub trait Trait: frame_system::Trait + BalancesTrait + IdentityTrait + asset::Trait {
+    /// The overarching event type.
+    type Event: From<Event> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// Weight information for extrinsics in the corporate actions pallet.
+    type WeightInfo: WeightInfo;
+}
+
+type Identity<T> = identity::Module<T>;
+type Asset<T> = asset::Module<T>;
+
+decl_storage! {
+    trait Store for Module<T: Trait> as CorporateAction {
+        /// A corporate action agent (CAA) of a ticker, if specified,
+        /// that may be different from the asset owner (AO).
+        /// If `None`, the AO is the CAA.
+        ///
+        /// The CAA may be distict from the AO because the CAA may require a money services license,
+        /// and the assets would need to originate from the CAA's identity, not the AO's identity.
+        ///
+        /// (ticker => did?)
+        pub Agent get(fn agent): map hasher(blake2_128_concat) Ticker => Option<IdentityId>;
+
+        /// The identities targeted by default for CAs for this ticker,
+        /// either to be excluded or included.
+        ///
+        /// (ticker => target identities)
+        pub DefaultTargetIdentities get(fn default_target_identities): map hasher(blake2_128_concat) Ticker => TargetIdentities;
+
+        /// The default amount of tax to withhold ("withholding tax", WT) for this ticker when distributing dividends.
+        ///
+        /// To understand withholding tax, for example, let's assume that you hold IKEA shares.
+        /// IKEA now decides to distribute 100 SEK to Alice.
+        /// Alice lives in Sweden, so Skatteverket (the Swedish tax authority) wants 30% of that.
+        /// Then those 100 * 30% are withheld from Alice, and IKEA will send them to Skatteverket.
+        ///
+        /// (ticker => % to withhold)
+        pub DefaultWitholdingTax get(fn default_withholding_tax): map hasher(blake2_128_concat) Ticker => Permill;
+
+        /// The amount of tax to withhold ("withholding tax", WT) for a certain ticker x DID.
+        /// If an entry exists for a certain DID, it overrides the default in `DefaultWithholdingTax`.
+        ///
+        /// (ticker => DID => % to withhold)
+        pub DidWitholdingTax get(fn did_withholding_tax):
+            double_map hasher(blake2_128_concat) Ticker, hasher(blake2_128_concat) IdentityId => Option<Permill>;
+    }
+}
+
+// Public interface for this runtime module.
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        /// initialize the default event for this module
+        fn deposit_event() = default;
+
+        /// Reset the CAA of `ticker` to its owner.
+        ///
+        /// ## Arguments
+        /// - `ticker` for which the CAA is reset.
+        ///
+        /// ## Errors
+        /// - `BadOrigin` if not signed
```

Comment:
I think we can skip listing `BadOrigin` in Errors. It's somewhat implicit and can only be triggered when a committee tries to call a function meant for general users.